modules/core/04-channel/keeper/upgrade.go

```
@@ -597,6 +597,94 @@ func (k Keeper) startFlushUpgradeHandshake(
+	}
+
+	var counterpartyChannel types.Channel
+	counterpartyHops := []string{connection.GetCounterparty().GetConnectionID()}
+	switch counterpartyChannelState {
+	case types.OPEN:
+		upgrade, found := k.GetUpgrade(ctx, portID, channelID)
+		if !found {
+			return errorsmod.Wrapf(types.ErrUpgradeNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+		}
+		counterpartyChannel = types.Channel{
+			State:           types.OPEN,
+			Ordering:        upgrade.Fields.Ordering,
+			ConnectionHops:  upgrade.Fields.ConnectionHops,
+			Counterparty:    types.NewCounterparty(portID, channelID),
+			Version:         upgrade.Fields.GetVersion(),
+			UpgradeSequence: channel.UpgradeSequence,
+			FlushStatus:     types.NOTINFLUSH,
+		}
+
+	case types.TRYUPGRADE:
+		// If the counterparty is in TRYUPGRADE, then we must have gone through the ACKUPGRADE step.
+		if channel.State != types.ACKUPGRADE {
+			return errorsmod.Wrapf(types.ErrInvalidChannelState, "expected %s, got %s", types.ACKUPGRADE, channel.State)
+		}
+		counterpartyChannel = types.Channel{
+			State:           types.TRYUPGRADE,
+			Ordering:        channel.Ordering,
+			ConnectionHops:  counterpartyHops,
+			Counterparty:    types.NewCounterparty(portID, channelID),
+			Version:         channel.Version,
+			UpgradeSequence: channel.UpgradeSequence,
+			FlushStatus:     types.FLUSHCOMPLETE,
+		}
+
+	case types.ACKUPGRADE:
+		counterpartyChannel = types.Channel{
+			State:           types.ACKUPGRADE,
+			Ordering:        channel.Ordering,
+			ConnectionHops:  counterpartyHops,
+			Counterparty:    types.NewCounterparty(portID, channelID),
+			Version:         channel.Version,
+			UpgradeSequence: channel.UpgradeSequence,
+			FlushStatus:     types.FLUSHCOMPLETE,
+		}
+	default:
+		panic(fmt.Sprintf("counterparty channel state should be in one of [%s, %s, %s]; got %s", types.TRYUPGRADE, types.ACKUPGRADE, types.OPEN, counterpartyChannelState))
```

I don't see any reason to panic when we can return an error?

---

mirrors related handling in `ChanUpgradeTry`. Guess panic indicates the unreachable-ness of it?